gammapy/scripts/downloadclasses.py

```diff
@@ -0,0 +1,316 @@
+# Licensed under a 3-clause BSD style license - see LICENSE.rst
+"""Download class for gammapy download CLI."""
+from __future__ import absolute_import, division, print_function, unicode_literals
+import hashlib
+import json
+import logging
+import sys
+import multiprocessing
+from ..extern.six.moves.urllib.request import urlretrieve, urlopen
+from ..extern.pathlib import Path
+from .. import version
+
+log = logging.getLogger(__name__)
+
+BASE_URL = "https://gammapy.org/download"
+DEV_NBS_YAML_URL = (
+    "https://raw.githubusercontent.com/gammapy/gammapy/master/tutorials/notebooks.yaml"
+)
+DEV_SCRIPTS_YAML_URL = (
+    "https://raw.githubusercontent.com/gammapy/gammapy/master/examples/scripts.yaml"
+)
+DEV_DATA_JSON_URL = (
+    "https://raw.githubusercontent.com/gammapy/gammapy-webpage/gh-pages/download/data/gammapy-data-index.json"
+)
+
+def hashmd5(path):
+    md5_hash = hashlib.md5()
+    with open(path, "rb") as f:
+        for byte_block in iter(lambda: f.read(4096), b""):
+            md5_hash.update(byte_block)
+    md5 = md5_hash.hexdigest()
+    return md5
+
+
+def get_file(ftuple):
+    url, filepath, md5server = ftuple
+
+    retrieve = True
+    if md5server and Path(filepath).exists():
+        md5local = hashmd5(filepath)
+        if md5local == md5server:
+            retrieve = False
+
+    if retrieve:
+        try:
+            urlretrieve(url, filepath)
+        except Exception as ex:
+            log.error(filepath + " could not be copied.")
+            log.error(ex)
+
+
+def parse_datafiles(datasearch, datasetslist):
+    datafiles = {}
+    for dataset in datasetslist:
+        if datasearch == dataset["name"] or datasearch == "":
+            if dataset["files"]:
+                for ds in dataset["files"]:
+                    label = ds["path"]
+                    datafiles[label] = {}
+                    datafiles[label]["url"] = ds["url"]
+                    datafiles[label]["path"] = ds["path"]
+                    if "hashmd5" in ds:
+                        datafiles[label]["hashmd5"] = ds["hashmd5"]
+    return datafiles
+
+
+def parse_imagefiles(notebookslist):
+    imagefiles = {}
```

Comment:
I think using generators ("yield") in the `parse_` functions would save a few lines and be a bit simpler.
Note that if you need lists or dicts instead of generators, you can always call `list` or `dict` on the return value from the generator function. It's not about efficiency in this case, but (minor) simpler / shorter code in the function making the list.